department-of-veterans-affairs/va.gov-cms
# VACMS-8014: Add Entity Usage module.

## Description
Adds Entity Usage to the platform.

What it does:
It tracks references across entities, and allows one to view (in a local task/tab) where the current entity is being used. This is helpful in content governance where a user might need to know where an entity is being used as a means to determine if un-publishing a node is possible based on what links to elsewhere in the CMS, for instance.

How it works:
Entity Usage maintains usage statistics in a dedicated database table, `entity_usage`. Ongoing tracking takes place during entity save operations. However, there is a one-time initial scan of all (configured) entities to seed the database. Because of the intense nature of scanning all nodes in the database, this process can take several hours or longer to complete.

Initial seeding:
Seeding can done either via a button on the Entity Usage settings form (/admin/config/entity-usage/settings), or via a Drush command.

PHP timeouts will occur if one attempts to use the UI for seeding the usage table, even despite the fact that it is using Drupal batch operations.

Fortunately the module ships with a Drush command for seeding the usage statistics `entity-usage:recreate`. However, using that command without options will also timeout after 4 hours (the PHP script timeout). The recommended approach is to use the `--use-queue` option when initially seeding. This will cause the queue to be populated with the items to track. This is a cron queue implementation with a 60 second timeout. Each cron run will process thousands of nodes (~11,000 locally). The queue will have a few million records once fully populated, so there could be hundreds of cron runs necessary to fully populate the usage statistics table.

Relates to #8014 .

## Testing done

## Screenshots
<img width="1836" alt="Screen Shot 2022-06-09 at 3 44 43 PM" src="https://user-images.githubusercontent.com/221539/172958105-56efc369-e949-4a73-9009-fdba760d252b.png">
<img width="2022" alt="Screen Shot 2022-06-09 at 3 45 46 PM" src="https://user-images.githubusercontent.com/221539/172958125-0d04f50f-6b5d-492a-9c57-32958ec9da8d.png">

## QA steps

As an editor
1. Edit this node /pension/aid-attendance-housebound
2. Re-save it as Published
3. Visit  /node/588/usage (this node is linked from the first node via a rich text field on a Paragraph reference field on the node.
4. Verify the node is being tracked (should look like screenshot 1 above)

### Definition of Done


### Select Team for PR review



### Is this PR blocked by another PR?


### Does this PR need review from a Product Owner


### CMS user-facing annoucement

Is an announcement needed to let editors know of this change?
